pkg/gui/branches_panel.go

```diff
@@ -9,154 +9,346 @@ import (
+			gui.Log.Errorf("Failed to create error panel at handleBranchPress: %s\n", err)
+			return err
+		}
+
+		return nil
 	}
-	branch := gui.getSelectedBranch(gui.getBranchesView(g))
-	if err := gui.GitCommand.Checkout(branch.Name, false); err != nil {
-		gui.createErrorPanel(g, err.Error())
+
+	branch := gui.getSelectedBranch(v)
+
+	err = gui.GitCommand.Checkout(branch.Name, false)
+	if err != nil {
+		err = gui.createErrorPanel(err.Error())
+		if err != nil {
+			gui.Log.Errorf("Failed to create error panel at handleBranchPress: %s\n", err)
+			return err
+		}
 	}
-	return gui.refreshSidePanels(g)
+
+	gui.refresh()
+
+	return nil
 }
 
+// handleForceCheckout is called when the user wants to force checkout a branch
+// g and v are passed by the gocui library, but are not used.
+// In case something goes wrong it returns an error
 func (gui *Gui) handleForceCheckout(g *gocui.Gui, v *gocui.View) error {
+
 	branch := gui.getSelectedBranch(v)
 	message := gui.Tr.SLocalize("SureForceCheckout")
 	title := gui.Tr.SLocalize("ForceCheckoutBranch")
-	return gui.createConfirmationPanel(g, v, title, message, func(g *gocui.Gui, v *gocui.View) error {
-		if err := gui.GitCommand.Checkout(branch.Name, true); err != nil {
-			gui.createErrorPanel(g, err.Error())
-		}
-		return gui.refreshSidePanels(g)
-	}, nil)
+
+	err := gui.createConfirmationPanel(v, title, message,
+		func(g *gocui.Gui, v *gocui.View) error {
+
+			err := gui.GitCommand.Checkout(branch.Name, true)
+			if err != nil {
+				err = gui.createErrorPanel(err.Error())
+				if err != nil {
+					gui.Log.Errorf("Failed to create error panel at handleForceCheckout: %s\n", err)
+					return err
+				}
+			}
+
+			gui.refresh()
+
+			return nil
+		}, nil)
+	if err != nil {
+		gui.Log.Errorf("Failed to create confirmation panel at handleForceCheckout: %s\n", err)
+		return err
+	}
+
+	return nil
 }
 
+// handleCheckoutByName gets called when a user presses the key
+// to checkout a branch by name.
+// g and v are passed by the gocui library, but only v is used.
+// If something goes wrong it returns an error
 func (gui *Gui) handleCheckoutByName(g *gocui.Gui, v *gocui.View) error {
-	gui.createPromptPanel(g, v, gui.Tr.SLocalize("BranchName")+":", func(g *gocui.Gui, v *gocui.View) error {
-		if err := gui.GitCommand.Checkout(gui.trimmedContent(v), false); err != nil {
-			return gui.createErrorPanel(g, err.Error())
-		}
-		return gui.refreshSidePanels(g)
-	})
+
+	err := gui.createPromptPanel(v, gui.Tr.SLocalize("BranchName")+":",
+		func(g *gocui.Gui, v *gocui.View) error {
+
+			err := gui.GitCommand.Checkout(gui.trimmedContent(v), false)
+			if err != nil {
+				err = gui.createErrorPanel(err.Error())
+				if err != nil {
+					gui.Log.Errorf("Failed to create error panel at handleCheckoutByName: %s\n", err)
+					return err
+				}
+
+				return nil
+			}
+
+			gui.refresh()
+
+			return nil
+		})
+	if err != nil {
+		gui.Log.Errorf("Failed to create prompt panel at handleCheckoutByName: %s\n", err)
```

Comment:
We seem to be sacrificing concision here for the sake of logging errors, but it's not clear to me that the tradeoff is worth it. Is it idiomatic to log an error before returning it, wherever we are in the code?

Reply:
Well since it is idiomatic to check every return value (check the bots output), and it can produce actually helpful feedback during crashes, i think we should always log before returning. As the error caused by the correctCursor function got solved by adding logs to the error checking, I think this is the only good way of finding errors quick. At last, if we send the errors to the web logger by default, we get every single tiny bug immediately in the inbox.

Reply:
I certainly agree that we should always handle errors, but bubbling up the error is sufficient, as we can log and then handle the error specifically at whatever level is appropriate to do that. We also have the issue that by logging at each level, we're going to be receiving error logs several times for the same error but with different messages, which muddies the water a bit in terms of error monitoring.

I agree that it can help with finding bugs, but I it sacrifices readability of the code too much, and prevents us from doing things like tail calls on functions that return errors. In the event that we need to do some debugging, I haven't so far had any difficulty adding logs as required when doing the debugging, because it's usually self-evident where to start in the debugging process.

We should bubble up errors and log them at the top level when we handle them in the code.